08-mmc-upload/mmc-upload.sh

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+SCRIPTPATH="$( cd "$( dirname "${BASH_SOURCE[0]}"  )" && pwd  )"
+
+# Source FPGA mapping
+. ${SCRIPTPATH}/../misc/crate-fpga-mapping.sh
+
+MMC_BINARIES_DIR_REL=binaries
+MMC_BINARIES_DIR=${SCRIPTPATH}/${MMC_BINARIES_DIR_REL}
+
+OPENMMC_VERSION="v1.6.0"
+
+AFCv3_BINARY="openMMC-afcv3.1-8sfp-${OPENMMC_VERSION}.hpm"
+AFCv4_BINARY="openMMC-afcv4.0-lamp-${OPENMMC_VERSION}.hpm"
+
+
+# Source common functions
+. ${SCRIPTPATH}/../misc/functions.sh
+
+set +u
+
+# Simple argument checking
+if [ $# -lt 4 ]; then
+    echo "Usage: ./mmc-upload.sh <MCH Hostname> <Crate Number> <Start Slot> <Commit Hash>"
+    exit 1
+fi
+
+MCH_IP="$1"
+CRATE_NUMBER="$2"
+START_SLOT="$3"
+COMMIT_HASH="$4"
+
+set -u
+
+set +e
+#Program binaries
+exec_cmd "INFO " echo "Programming MMCs..."
+BPM_MAX_NUM_BOARDS=12
+for slot in `seq ${START_SLOT} ${BPM_MAX_NUM_BOARDS}`; do
+
+  bpm_crate_fpga_raw="BPM_CRATE_${CRATE_NUMBER}_FPGA[${slot}]"
+  bpm_crate_fpga=${!bpm_crate_fpga_raw}
+
+  case ${bpm_crate_fpga} in
+    "timing")
+      binary=${MMC_BINARIES_DIR}/${AFCv3_BINARY}
+    ;;
+    "fofb")
+      binary=${MMC_BINARIES_DIR}/${AFCv4_BINARY}
+    ;;
+    "pbpm")
+      binary=${MMC_BINARIES_DIR}/${AFCv3_BINARY}
+    ;;
+    "sr")
+      binary=${MMC_BINARIES_DIR}/${AFCv3_BINARY}
+    ;;
+    "bo")
+      binary=${MMC_BINARIES_DIR}/${AFCv3_BINARY}
+    ;;
+    "")
+      # Skip it
+      continue
+    ;;
+    *)
+      exec_cmd "ERR  " echo "Invalid bitstream type: ${bpm_crate_fpga}" >&2
+      exit 1
+    ;;
+  esac
+
+  tries=0
+
+  echo y | ~/sw/ipmitool/bin/ipmitool -I lan -H ${MCH_IP} -A  none -T 0x82 -m 0x20 -t $((112 + 2*${slot})) hpm upgrade ${binary} activate
```

Comment:
This assumes a particular path for ipmitool. Change it to use the version in $PATH (aka don't hard-code the ipmitool path). If you want to use a different ipmitool binary, the user is responsible in inserting the path corresponding path at the beginning of the PATH variable before calling this script.